File: JavaExercise/src/top/interview/Design/ShuffleAnArray.java
package top.interview.Design;

import java.util.*;
import java.util.stream.Collectors;

/**
 * https://leetcode.com/explore/interview/card/top-interview-questions-easy/98/design/670/
 */
public class ShuffleAnArray {
    public int[] array;
    public ShuffleAnArray(int[] nums) {
        array = nums;
    }

    /** Resets the array to its original configuration and return it. */
    public int[] reset() {
        return array;
    }

    /** Returns a random shuffling of the array. */
    public int[] shuffle() {
        List<Integer> llist = Arrays.stream(array).boxed().collect(Collectors.toList());
        Collections.shuffle(llist);
        return llist.stream().mapToInt(Integer::valueOf).toArray();
    }

    public static void main(String[] args) {
        int[] nums = {1,2,3};
        ShuffleAnArray shuffleAnArray = new ShuffleAnArray(nums);
        int[] res1 = shuffleAnArray.reset();
        int[] res2 = shuffleAnArray.shuffle();
    }
}

File: tencent/shuffle-an-array.py
# -*- coding: utf-8 -*-
"""

   File Name：     shuffle-an-array
   Author :       jing
   Date：          2020/3/25

   https://leetcode-cn.com/problems/shuffle-an-array/

   打乱数组
"""


class Solution:

    def __init__(self, nums):
        self.nums = nums
        self.copy = nums[:]

    def reset(self):
        """
        Resets the array to its original configuration and return it.
        """
        return self.copy

    def shuffle(self):
        """
        Returns a random shuffling of the array.
        """
        # import random
        # random.shuffle(self.nums)
        # return self.nums
        import random
        for i in range(len(self.nums)):
            j = random.randint(i, len(self.nums) - 1)
            self.nums[i], self.nums[j] = self.nums[j], self.nums[i]
        return self.nums
